File: python/medousa-sdk/src/medousa/sync/agents.py
from __future__ import annotations

from medousa._decode import decode
from medousa._ops import op_path, op_path_query
from medousa.sync.client import MedousaClientSync
from medousa.types import (
    AgentPermissionRequestListResponse,
    AgentPermissionResolveRequest,
    AgentPermissionResolveResponse,
    AgentRuntimeListResponse,
    AgentSecretRequestListResponse,
    AgentSecretResolveResponse,
    AgentSessionPromptRequest,
    AgentSessionPromptResponse,
    CancelAgentSessionResponse,
    CreateAgentSessionRequest,
    CreateAgentSessionResponse,
)


class AgentsApiSync:
    def __init__(self, client: MedousaClientSync) -> None:
        self._client = client

    def list_runtimes(self) -> AgentRuntimeListResponse:
        value = self._client._transport.get_json(
            self._client.base_url, op_path("agents.runtimes.get")
        )
        return decode(AgentRuntimeListResponse, value)

    def create_session(self, request: CreateAgentSessionRequest) -> CreateAgentSessionResponse:
        value = self._client._transport.post_json(
            self._client.base_url,
            op_path("agents.sessions.post"),
            request.model_dump(mode="json", exclude_none=True),
        )
        return decode(CreateAgentSessionResponse, value)

    def prompt(
        self, agent_session_id: str, request: AgentSessionPromptRequest
    ) -> AgentSessionPromptResponse:
        value = self._client._transport.post_json(
            self._client.base_url,
            op_path(
                "agents.sessions.by_agent_session_id.prompt.post",
                agent_session_id=agent_session_id.strip(),
            ),
            request.model_dump(mode="json", exclude_none=True),
        )
        return decode(AgentSessionPromptResponse, value)

    def cancel(self, agent_session_id: str) -> CancelAgentSessionResponse:
        value = self._client._transport.post_empty_json(
            self._client.base_url,
            op_path(
                "agents.sessions.by_agent_session_id.cancel.post",
                agent_session_id=agent_session_id.strip(),
            ),
        )
        return decode(CancelAgentSessionResponse, value)

    def list_permission_requests(
        self, *, status: str | None = "pending", limit: int | None = None
    ) -> AgentPermissionRequestListResponse:
        query: list[tuple[str, str]] = []
        if status is not None:
            query.append(("status", status))
        if limit is not None:
            query.append(("limit", str(limit)))
        route = op_path_query("agents.permission_requests.get", query)
        value = self._client._transport.get_json(self._client.base_url, route)
        return decode(AgentPermissionRequestListResponse, value)

    def approve_permission(
        self, request_id: str, request: AgentPermissionResolveRequest | None = None
    ) -> AgentPermissionResolveResponse:
        body = (request or AgentPermissionResolveRequest()).model_dump(
            mode="json", exclude_none=True
        )
        value = self._client._transport.post_json(
            self._client.base_url,
            op_path(
                "agents.permission_requests.by_request_id.approve.post",
                request_id=request_id.strip(),
            ),
            body,
        )
        return decode(AgentPermissionResolveResponse, value)

    def deny_permission(
        self, request_id: str, request: AgentPermissionResolveRequest | None = None
    ) -> AgentPermissionResolveResponse:
        body = (request or AgentPermissionResolveRequest()).model_dump(
            mode="json", exclude_none=True
        )
        value = self._client._transport.post_json(
            self._client.base_url,
            op_path(
                "agents.permission_requests.by_request_id.deny.post", request_id=request_id.strip()
            ),
            body,
        )
        return decode(AgentPermissionResolveResponse, value)

    def list_secret_requests(
        self, *, status: str | None = "pending", limit: int | None = None
    ) -> AgentSecretRequestListResponse:
        query: list[tuple[str, str]] = []
        if status is not None:
            query.append(("status", status))
        if limit is not None:
            query.append(("limit", str(limit)))
        route = op_path_query("agents.secret_requests.get", query)
        value = self._client._transport.get_json(self._client.base_url, route)
        return decode(AgentSecretRequestListResponse, value)

    def fulfill_secret_request(
        self, request_id: str, value: str, *, resolved_by: str | None = None
    ) -> AgentSecretResolveResponse:
        """Move a write-only value directly to the request's workshop backend."""
        body = {"value": value}
        if resolved_by is not None:
            body["resolved_by"] = resolved_by
        response = self._client._transport.post_json(
            self._client.base_url,
            op_path(
                "agents.secret_requests.by_request_id.fulfill.post",
                request_id=request_id.strip(),
            ),
            body,
        )
        return decode(AgentSecretResolveResponse, response)

    def deny_secret_request(
        self, request_id: str, *, resolved_by: str | None = None
    ) -> AgentSecretResolveResponse:
        body = {} if resolved_by is None else {"resolved_by": resolved_by}
        value = self._client._transport.post_json(
            self._client.base_url,
            op_path(
                "agents.secret_requests.by_request_id.deny.post",
                request_id=request_id.strip(),
            ),
            body,
        )
        return decode(AgentSecretResolveResponse, value)

File: crates/medousa-sdk/src/agents.rs
#[cfg(feature = "async")]
use medousa_types::{
    AgentPermissionRequestListResponse, AgentPermissionResolveRequest,
    AgentPermissionResolveResponse, AgentRuntimeListResponse, AgentSecretDenyRequest,
    AgentSecretFulfillRequest, AgentSecretRequestListResponse, AgentSecretResolveResponse,
    AgentSessionPromptRequest, AgentSessionPromptResponse, CancelAgentSessionResponse,
    CreateAgentSessionRequest, CreateAgentSessionResponse, InteractiveTurnStreamEvent,
    SetAgentSessionConfigOptionRequest, SetAgentSessionConfigOptionResponse,
};

#[cfg(all(feature = "async", feature = "sse"))]
use futures_util::Stream;
#[cfg(all(feature = "async", feature = "sse"))]
use futures_util::StreamExt;

#[cfg(feature = "async")]
use crate::client::MedousaClient;
use crate::generated::ops;
use crate::op::{op_path, op_path_query};
#[cfg(feature = "async")]
use crate::transport::decode;

#[cfg(all(feature = "async", feature = "sse"))]
use crate::streaming::{SseLineStream, decode_sse_json};

#[cfg(feature = "async")]
pub struct AgentsApi<'a> {
    pub(crate) client: &'a MedousaClient,
}

#[cfg(feature = "async")]
impl AgentsApi<'_> {
    pub async fn list_runtimes(&self) -> Result<AgentRuntimeListResponse, crate::SdkError> {
        let value = self
            .client
            .transport()
            .get_json(self.client.base_url(), ops::AGENTS_RUNTIMES_GET.path)
            .await?;
        decode(value).await
    }

    pub async fn create_session(
        &self,
        request: &CreateAgentSessionRequest,
    ) -> Result<CreateAgentSessionResponse, crate::SdkError> {
        let body =
            serde_json::to_value(request).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), ops::AGENTS_SESSIONS_POST.path, body)
            .await?;
        decode(value).await
    }

    pub async fn prompt(
        &self,
        agent_session_id: &str,
        request: &AgentSessionPromptRequest,
    ) -> Result<AgentSessionPromptResponse, crate::SdkError> {
        let body =
            serde_json::to_value(request).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let path = op_path(
            &ops::AGENTS_SESSIONS_BY_AGENT_SESSION_ID_PROMPT_POST,
            &[("agent_session_id", agent_session_id.trim())],
        )?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), &path, body)
            .await?;
        decode(value).await
    }

    pub async fn set_config_option(
        &self,
        agent_session_id: &str,
        request: &SetAgentSessionConfigOptionRequest,
    ) -> Result<SetAgentSessionConfigOptionResponse, crate::SdkError> {
        let body =
            serde_json::to_value(request).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let path = op_path(
            &ops::AGENTS_SESSIONS_BY_AGENT_SESSION_ID_CONFIG_POST,
            &[("agent_session_id", agent_session_id.trim())],
        )?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), &path, body)
            .await?;
        decode(value).await
    }

    pub async fn cancel(
        &self,
        agent_session_id: &str,
    ) -> Result<CancelAgentSessionResponse, crate::SdkError> {
        let path = op_path(
            &ops::AGENTS_SESSIONS_BY_AGENT_SESSION_ID_CANCEL_POST,
            &[("agent_session_id", agent_session_id.trim())],
        )?;
        let value = self
            .client
            .transport()
            .post_empty_json(self.client.base_url(), &path)
            .await?;
        decode(value).await
    }

    #[cfg(feature = "sse")]
    pub fn stream(
        &self,
        stream_url: impl Into<String>,
    ) -> impl Stream<Item = Result<InteractiveTurnStreamEvent, crate::SdkError>> + '_ {
        let byte_stream = self
            .client
            .transport()
            .stream_sse(self.client.base_url(), stream_url.into());
        SseLineStream::new(byte_stream).map(|line| line.and_then(|data| decode_sse_json(&data)))
    }

    #[cfg(feature = "sse")]
    pub async fn stream_session(
        &self,
        request: &CreateAgentSessionRequest,
    ) -> Result<
        impl Stream<Item = Result<InteractiveTurnStreamEvent, crate::SdkError>> + '_,
        crate::SdkError,
    > {
        let response = self.create_session(request).await?;
        Ok(self.stream(response.stream_url))
    }

    pub async fn list_permission_requests(
        &self,
        status: Option<&str>,
        limit: Option<usize>,
    ) -> Result<AgentPermissionRequestListResponse, crate::SdkError> {
        let mut params = Vec::new();
        if let Some(status) = status {
            params.push(("status", status.to_string()));
        }
        if let Some(limit) = limit {
            params.push(("limit", limit.to_string()));
        }
        let path = op_path_query(&ops::AGENTS_PERMISSION_REQUESTS_GET, &[], &params)?;
        let value = self
            .client
            .transport()
            .get_json(self.client.base_url(), &path)
            .await?;
        decode(value).await
    }

    pub async fn approve_permission(
        &self,
        request_id: &str,
        request: &AgentPermissionResolveRequest,
    ) -> Result<AgentPermissionResolveResponse, crate::SdkError> {
        let body =
            serde_json::to_value(request).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let path = op_path(
            &ops::AGENTS_PERMISSION_REQUESTS_BY_REQUEST_ID_APPROVE_POST,
            &[("request_id", request_id.trim())],
        )?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), &path, body)
            .await?;
        decode(value).await
    }

    pub async fn deny_permission(
        &self,
        request_id: &str,
        request: &AgentPermissionResolveRequest,
    ) -> Result<AgentPermissionResolveResponse, crate::SdkError> {
        let body =
            serde_json::to_value(request).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let path = op_path(
            &ops::AGENTS_PERMISSION_REQUESTS_BY_REQUEST_ID_DENY_POST,
            &[("request_id", request_id.trim())],
        )?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), &path, body)
            .await?;
        decode(value).await
    }

    pub async fn list_secret_requests(
        &self,
        status: Option<&str>,
        limit: Option<usize>,
    ) -> Result<AgentSecretRequestListResponse, crate::SdkError> {
        let mut params = Vec::new();
        if let Some(status) = status {
            params.push(("status", status.to_string()));
        }
        if let Some(limit) = limit {
            params.push(("limit", limit.to_string()));
        }
        let path = op_path_query(&ops::AGENTS_SECRET_REQUESTS_GET, &[], &params)?;
        let value = self
            .client
            .transport()
            .get_json(self.client.base_url(), &path)
            .await?;
        decode(value).await
    }

    pub async fn fulfill_secret_request(
        &self,
        request_id: &str,
        request: &AgentSecretFulfillRequest,
    ) -> Result<AgentSecretResolveResponse, crate::SdkError> {
        let body =
            serde_json::to_value(request).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let path = op_path(
            &ops::AGENTS_SECRET_REQUESTS_BY_REQUEST_ID_FULFILL_POST,
            &[("request_id", request_id.trim())],
        )?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), &path, body)
            .await?;
        decode(value).await
    }

    pub async fn deny_secret_request(
        &self,
        request_id: &str,
        request: &AgentSecretDenyRequest,
    ) -> Result<AgentSecretResolveResponse, crate::SdkError> {
        let body =
            serde_json::to_value(request).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let path = op_path(
            &ops::AGENTS_SECRET_REQUESTS_BY_REQUEST_ID_DENY_POST,
            &[("request_id", request_id.trim())],
        )?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), &path, body)
            .await?;
        decode(value).await
    }
}
